python/cudf/cudf/core/dataframe.py

```diff
@@ -2750,92 +2731,78 @@ def set_index(
+                # Try coerce into column
+                if not is_column_like(col):
+                    try:
+                        col = as_column(col)
+                    except TypeError:
+                        raise f"{col} cannot be converted to column-like."
+                if isinstance(col, (cudf.MultiIndex, pd.MultiIndex)):
+                    col = (
+                        cudf.from_pandas(col)
+                        if isinstance(col, pd.MultiIndex)
+                        else col
+                    )
+                    cols = [col._data[x] for x in col._data]
+                    columns_to_add = columns_to_add + cols
+                    names = names + col.names
+                else:
+                    # Column-like
+                    columns_to_add.append(col)
+                    if isinstance(
+                        col, (cudf.Series, cudf.Index, pd.Series, pd.Index)
+                    ):
+                        names.append(col.name)
+                    else:
+                        names.append(None)
 
-        # When index is a column name
-        if isinstance(index, str):
-            idf = self[index]
-            names = [index]
-            if append:
-                idf, names = _merge_index_frame(idf, names)
-                index = cudf.MultiIndex.from_frame(idf, names=names)
-            else:
-                index = as_index(idf)
-            return self._set_index(
-                index=index,
-                inplace=inplace,
-                drop=drop,
-                append=append,
-                verify_integrity=verify_integrity,
+        if len(col_not_found) > 0:
+            raise ValueError(
+                f"The following columns are not found: {col_not_found}"
             )
 
-        # When index is a new `Index` or `Series-convertible`
-        # Will not drop columns
-        index = index if isinstance(index, Index) else as_index(index)
-        idf = DataFrame(index._data)
-        names = (
-            index.names if isinstance(index, cudf.MultiIndex) else [index.name]
-        )
         if append:
-            idf, names = _merge_index_frame(idf, names)
-            index = cudf.MultiIndex.from_frame(idf, names=names)
+            idx_cols = [self.index._data[x] for x in self.index._data]
+            if isinstance(self.index, cudf.MultiIndex):
+                idx_names = self.index.names
+            else:
+                idx_names = [self.index.name]
+            columns_to_add = idx_cols + columns_to_add
+            names = idx_names + names
+
+        if len(columns_to_add) == 0:
+            raise ValueError("No valid columns to be added to index.")
+        elif len(columns_to_add) == 1:
+            idx = cudf.Index(columns_to_add[0], name=names[0])
+        else:
+            idf = cudf.DataFrame()
+            for i, col in enumerate(columns_to_add):
+                idf[i] = col
+            idx = cudf.MultiIndex(source_data=idf, names=names)
```

Comment:
Instead of using the `source_data` constructor here can we use `cudf.MultiIndex.from_frame`? We're trying to get rid of `source_data` in the future.